goal_src/levels/common/battlecontroller.gc

```diff
@@ -5,3 +5,1043 @@
+                                      (else
+                                       (set!
+                                        (-> v1-4 fact pickup-type)
+                                        (the-as
+                                         pickup-type
+                                         (-> self final-pickup-type)
+                                         )
+                                        )
+                                       (set! (-> v1-4 fact pickup-amount) 1.0)
+                                       (let ((f0-1 0.0))
+                                        (set!
+                                         (-> v1-4 fact pickup-spawn-amount)
+                                         f0-1
+                                         )
+                                        f0-1
+                                        )
+                                       )
+                                      )
+                         )
+                        )
+                       )
+                      )
+                     )
+                    )
+                   ((= v1-0 'trigger)
+                    (go-virtual battlecontroller-die)
+                    )
+                   )
+    )
+   )
+  )
+
+battlecontroller-default-event-handler
+
+(defbehavior battlecontroller-draw-debug battlecontroller ()
+  (dotimes (gp-0 (-> self spawner-count))
+   (dummy-9 (-> self spawner-array gp-0 path))
+   )
+  0
+  (none)
+  )
+
+(defbehavior battlecontroller-camera-on battlecontroller ()
+  (when (not (-> self camera-on))
+   (set! (-> self camera-on) #t)
+   (let ((v1-4 (res-lump-struct (-> self entity) 'camera-name structure)))
+    (when v1-4
+     (let ((a1-1 (new 'stack-no-clear 'event-message-block)))
+      (set! (-> a1-1 from) self)
+      (set! (-> a1-1 num-params) 1)
+      (set! (-> a1-1 message) 'change-to-entity-by-name)
+      (set! (-> a1-1 param 0) (the-as uint v1-4))
+      (send-event-function *camera* a1-1)
+      )
+     )
+    )
+   )
+  (the-as object 0)
+  )
+
+(defbehavior battlecontroller-camera-off battlecontroller ()
+  (set! (-> self camera-on) #f)
+  (let ((a1-0 (new 'stack-no-clear 'event-message-block)))
+   (set! (-> a1-0 from) self)
+   (set! (-> a1-0 num-params) 0)
+   (set! (-> a1-0 message) 'clear-entity)
+   (send-event-function *camera* a1-0)
+   )
+  0
+  (none)
+  )
+
+(defbehavior battlecontroller-update-spawners battlecontroller ()
+  (dotimes (gp-0 (-> self spawner-count))
+   (let* ((s5-0 (-> self spawner-array gp-0))
+          (s4-0 (handle->process (-> s5-0 creature)))
+          )
+    (when
+     (and s4-0 (logtest? (-> (the-as nav-enemy s4-0) nav-enemy-flags) 2048))
+     (cond
+      ((< (-> s5-0 state) (-> s5-0 path curve num-cverts))
+       (when (or (-> self noticed-player) (= (-> s5-0 state) 1))
+        (let ((s3-0 (new 'stack-no-clear 'vector)))
+         (eval-path-curve-div!
+          (-> s5-0 path)
+          s3-0
+          (the float (-> s5-0 state))
```

Comment:
`eval-path-curve-div!` should probably take an `int` for this argument, but we don't have to worry about that now

Reply:
hmm are you sure, it seems to explicitly convert to a float?
```asm
    mtc1 f0, s5               ;; [ 20] (set! f0-0 (gpr->fpr s5-0)) [s5: int ] -> []
    cvt.s.w f0, f0            ;; [ 21] (set! f0-1 (i2f f0-0)) [] -> []
    mfc1 a2, f0               ;; [ 22] (set! a2-0 (fpr->gpr f0-1)) [] -> [a2: float ]
```

Reply:
oops, you're right - it should definitely stay the way it is. I read it as `(the-as float` instead.